### PR TITLE
[Core] Adding compatibility with flags for VTKProcess

### PR DESCRIPTION
**Pull request by loumalouomega**
This adds compatibility for flags on nodes/elements and conditions in the VTKProcess. This adds too the possibility top postprocess bool variables (added to the communicator too)

The corresponding test has been added

### CODE REVIEW

**Comment by loumalouomega** [comment]
There are many files modified because of the many files that are for the vtk_process

File: kratos/input_output/vtk_output.cpp
Comment (philbucher) [inline kratos/input_output/vtk_output.cpp]: @adityaghantasala and I discussed this before and agreed on the following:

The value should be
-1 for undefined flags
0 for defined flags that are false
1 for defined flags that are true

File: kratos/python_scripts/vtk_output_process.py
Comment (philbucher) [inline kratos/python_scripts/vtk_output_process.py]: Flags

**Review by philbucher — CHANGES_REQUESTED** [review]
Thanks for the effort!
I am not sure if this can already go in, @adityaghantasala and I wanted to wait until the synchronization of flags is merged to implement this.
We will speak with @jcotela next week, but afaik this should happen soon
Otherwise it will not work properly in MPI and we might have to change things later

Also some minor comments related to the behavior

**Comment by philbucher** [comment]
Btw @adityaghantasala did you check the gp-output in MPI?
I don’t think anyone used the process from the MeshingApp in MPI before, I could imagine that some synchronization-calls are missing
=> we should check

File: kratos/input_output/vtk_output.cpp
Comment (loumalouomega) [reply kratos/input_output/vtk_output.cpp]: OK

**Comment by jcotela** [comment]
Regarding the flags issue: Synchronization for nodal `Kratos::Flags` (not `Variable<bool>` data) is implemented in  #3672 and should be available soon. We could use a similar approach for `bool` variables, but it does not exist at the moment.

**Comment by loumalouomega** [comment]
> Regarding the flags issue: Synchronization for nodal `Kratos::Flags` (not `Variable<bool>` data) is implemented in #3672 and should be available soon. We could use a similar approach for `bool` variables, but it does not exist at the moment.

OK

**Comment by philbucher** [comment]
@loumalouomega check codacy, I think there is an error in your test

**Comment by loumalouomega** [comment]
True

**Comment by loumalouomega** [comment]
Corrected

File: kratos/tests/test_vtk_output_process.py
Comment (philbucher) [inline kratos/tests/test_vtk_output_process.py]:
```suggestion
        cond.Set(KratosMultiphysics.BOUNDARY)
```

File: kratos/tests/vtk_output_process_ref_files/ascii3D/Main_MovingNodes_0_6.vtk
Comment (philbucher) [inline kratos/tests/vtk_output_process_ref_files/ascii3D/Main_MovingNodes_0_6.vtk]: why do those change?

Comment (adityaghantasala) [reply kratos/tests/vtk_output_process_ref_files/ascii3D/Main_MovingNodes_0_6.vtk]: This changed because now there are three data on the cells DENSITY, YOUNGS_MOD and BOUNDARY

File: kratos/tests/test_vtk_output_process.py
Comment (loumalouomega) [reply kratos/tests/test_vtk_output_process.py]: Ups

File: kratos/tests/vtk_output_process_ref_files/ascii3D/Main_MovingNodes_0_6.vtk
Comment (loumalouomega) [reply kratos/tests/vtk_output_process_ref_files/ascii3D/Main_MovingNodes_0_6.vtk]: Because the flag is one more variable

Comment (philbucher) [reply kratos/tests/vtk_output_process_ref_files/ascii3D/Main_MovingNodes_0_6.vtk]: ok thx

File: kratos/tests/test_vtk_output_process.py
Comment (loumalouomega) [reply kratos/tests/test_vtk_output_process.py]: Done

**Comment by adityaghantasala** [comment]
@loumalouomega  can you add this as reference in the cpp file header ?

https://vtk.org/wp-content/uploads/2015/04/file-formats.pdf

Also why are there changes to the communicator in this PR ?

**Comment by loumalouomega** [comment]
> @loumalouomega can you add this as reference in the cpp file header ?
> 
> [vtk.org/wp-content/uploads/2015/04/file-formats.pdf](https://vtk.org/wp-content/uploads/2015/04/file-formats.pdf)
> 
> Also why are there changes to the communicator in this PR ?

I can revert it, but are basically to be able to compile with Variable<bool>

**Comment by loumalouomega** [comment]
> @loumalouomega can you add this as reference in the cpp file header ?
> 
> [vtk.org/wp-content/uploads/2015/04/file-formats.pdf](https://vtk.org/wp-content/uploads/2015/04/file-formats.pdf)
This is already in the header

File: kratos/includes/communicator.h
Comment (loumalouomega) [inline kratos/includes/communicator.h]: I am just adding the communicator for the bool

Comment (loumalouomega) [reply kratos/includes/communicator.h]: @adityaghantasala

Comment (jcotela) [reply kratos/includes/communicator.h]: Note that this does not exist for the MPI communicator, so whatever it is that needs this method will not work in MPI.

Comment (adityaghantasala) [reply kratos/includes/communicator.h]: @loumalouomega  considering @jcotela 's comment does it help adding those changes here ? 
I really will prefer to have them in a different PR !

Comment (loumalouomega) [reply kratos/includes/communicator.h]: I added to MPI too, maybe the implementation is wrong, depending how do we consider the bool works in paralell

Comment (jcotela) [reply kratos/includes/communicator.h]: @loumalouomega does it compile?

Comment (loumalouomega) [reply kratos/includes/communicator.h]: Indeed, but I will create a separated PR, so you can add feedback directly

Comment (loumalouomega) [reply kratos/includes/communicator.h]: I have check and I had deactivated the MPI compilation, one sec

Comment (loumalouomega) [reply kratos/includes/communicator.h]: I have compilation errors with Metis application, I will post an issue

**Comment by loumalouomega** [comment]
@jcotela #3672  is merged, what should I do in this PR?

**Comment by jcotela** [comment]
@loumalouomega #3672 does not affect this PR directly, since that one deals with nodal `Flags` and here we are concerned about `Variable<bool>`.

However, there is something that I would like to ask you:

Could you please add a test for the synchronization of bool variables in #4571? I am already upgrading the MPICommunicator on a different branch and I could use the opportunity to add a dedicated method for the Synchronization you need. Having a working test would make it easier for me to check the new implementation.

**Comment by loumalouomega** [comment]
> @loumalouomega #3672 does not affect this PR directly, since that one deals with nodal `Flags` and here we are concerned about `Variable<bool>`.
> 
> However, there is something that I would like to ask you:
> 
> Could you please add a test for the synchronization of bool variables in #4571? I am already upgrading the MPICommunicator on a different branch and I could use the opportunity to add a dedicated method for the Synchronization you need. Having a working test would make it easier for me to check the new implementation.

:+1:

**Comment by loumalouomega** [comment]
Ooooh god, so many conflicts!!!

**Comment by jcotela** [comment]
@loumalouomega the MPI functions you require for bool should be in master now. I believe that you can just take the entire mpi_communicator.h from master to solve the merge conflict and the functions will be there. Please let me know if you encounter any trouble with them.

**Comment by loumalouomega** [comment]
Thanks @jcotela , can we merge now?

**Comment by jcotela** [comment]
@loumalouomega I do not have further comments, but I leave the last word to @philbucher and @adityaghantasala.

**Comment by loumalouomega** [comment]
:+1:

**Comment by philbucher** [comment]
There are still changes in the communicator

I will review tmr

File: kratos/input_output/vtk_output.cpp
Comment (philbucher) [inline kratos/input_output/vtk_output.cpp:459]: no, @adityaghantasala explained to me that we have no control over what paraview will display in case there is an element and a condition in the same location

We were discussing of introducing the "WriteConditions" flag same as the GiD output has

File: kratos/input_output/vtk_output.cpp
Comment (philbucher) [inline kratos/input_output/vtk_output.cpp:475]: just to say that we will only be able to write Flags that are defined in the Core
App-Flags will not be able to be written since they are not registered.

@jcotela am I wrong here?

File: kratos/input_output/vtk_output.cpp
Comment (philbucher) [inline kratos/input_output/vtk_output.cpp]: I am pretty convinced that those should be the local elements and not all of them

File: kratos/input_output/vtk_output.cpp
Comment (philbucher) [inline kratos/input_output/vtk_output.cpp]: same

File: kratos/input_output/vtk_output.cpp
Comment (philbucher) [inline kratos/input_output/vtk_output.cpp]: this is implemented now

@jcotela should it be `SynchronizeOrNodalFlags` or `SynchronizeAndNodalFlags`? I am not sure

File: kratos/mpi/includes/mpi_communicator.h
Comment (philbucher) [inline kratos/mpi/includes/mpi_communicator.h]: this should also be removed, I am quite sure it breaks the MPI-compilation

**Review by philbucher — COMMENTED** [review]
please remove the changes in the Communcator and implement my comments, then i will test it in MPI to make sure it also works there
Thanks!

File: kratos/input_output/vtk_output.cpp
Comment (philbucher) [reply kratos/input_output/vtk_output.cpp:459]: => this we can start doing after this is merged, then we can also get rid of the annoying warning

Other than that we also spoke abt a "WriteDeformedConfiguration" flag, same as GiD has

File: kratos/input_output/vtk_output.cpp
Comment (loumalouomega) [reply kratos/input_output/vtk_output.cpp:475]: If you are loading the applications other flags should appear

Comment (philbucher) [reply kratos/input_output/vtk_output.cpp:475]: hm I thought the same but I tried this in the IGAApp and there we could not retrieve the local iga flags from the Globals 
I can try again though in the scope of this PR

I mean, it does not affect this PR, there is not much we can do if it is not working

Comment (jcotela) [reply kratos/input_output/vtk_output.cpp:475]: I understand that application flags should be available as @loumalouomega says (but I never tested it). What we won't have is class-specific flags (such as those defined by the IO class or the TetrahedralMeshOrientationCheck, for example), which are not registered.

File: kratos/input_output/vtk_output.cpp
Comment (jcotela) [reply kratos/input_output/vtk_output.cpp]: I think that one normally wants SynchronizeOr (flag will be set to true if it is true in any of the local copies of the node), but this could ultimately be a case-specific choice.
Maybe I should add a Synchronize that sets (or unsets) it to whatever is the flag value in the owner process?

File: kratos/mpi/includes/mpi_communicator.h
Comment (jcotela) [reply kratos/mpi/includes/mpi_communicator.h]: Yes this is the old version... Probably the merge went wrong, an updated version of this function already exists in master.

File: kratos/input_output/vtk_output.cpp
Comment (loumalouomega) [reply kratos/input_output/vtk_output.cpp]: Done

File: kratos/input_output/vtk_output.cpp
Comment (loumalouomega) [reply kratos/input_output/vtk_output.cpp]: Done

**Comment by loumalouomega** [comment]
Done @philbucher

File: kratos/mpi/includes/mpi_communicator.h
Comment (loumalouomega) [reply kratos/mpi/includes/mpi_communicator.h]: This is outdated, I removed it from the files (I restored master versions)

**Comment by philbucher** [comment]
Thanks, changes look ok now, just need a little time to test it in MPI

**Comment by loumalouomega** [comment]
@philbucher, could you check this?

**Comment by philbucher** [comment]
now it works, requires #4901 to compile

**Review by philbucher — APPROVED** [review]
this will compile and work after #4901 is merged

**Comment by loumalouomega** [comment]
@philbucher this is ready then?

**Comment by philbucher** [comment]
> @philbucher this is ready then?

yes, please go ahead
thanks :+1:

**Comment by loumalouomega** [comment]
OK, then

**Pull request merged**
